pkg/operation/botanist/component/vpa/vpa.go

```diff
@@ -0,0 +1,457 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpa
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/kubeapiserver"
+	"github.com/gardener/gardener/pkg/utils"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// New creates a new instance of DeployWaiter for the Kubernetes Vertical Pod Autoscaler.
+func New(
+	client client.Client,
+	namespace string,
+	secretsManager secretsmanager.Interface,
+	values Values,
+) component.DeployWaiter {
+	v := &vpa{
+		client:         client,
+		namespace:      namespace,
+		secretsManager: secretsManager,
+		values:         values,
+	}
+
+	if values.ClusterType == ClusterTypeSeed {
+		v.registry = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+	} else {
+		v.registry = managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
+		v.crdDeployer = NewCRD(nil, v.registry)
+	}
+
+	return v
+}
+
+type vpa struct {
+	client         client.Client
+	namespace      string
+	secretsManager secretsmanager.Interface
+	values         Values
+
+	registry    *managedresources.Registry
+	crdDeployer component.Deployer
+
+	caSecretName     string
+	caBundle         []byte
+	serverSecretName string
+}
+
+// Values is a set of configuration values for the VPA components.
+type Values struct {
+	// ClusterType specifies the type of the cluster to which VPA is being deployed.
+	// For seeds, all resources are being deployed as part of a ManagedResource (except for the CRDs - those must be
+	// deployed separately because the VPA components themselves create VPA resources, hence the CRD must exist
+	// beforehand).
```

Comment:
Where's this CRDs deployed?

Reply:
Ah, I see it. 
https://github.com/gardener/gardener/blob/a548eb9c6399e0736301f816126d4e9f077eec61/pkg/operation/seed/seed.go#L445-L449